[AAE-10314] Fixed unhandled promises in ADF Core

Venrofi opened the pull request:
**Please check if the PR fulfills these requirements**

> - [x] The commit message follows our [guidelines](https://github.com/Alfresco/alfresco-ng2-components/wiki/Commit-format)
> - [ ] Tests for the changes have been added (for bug fixes / features)
> - [ ] Docs have been added / updated (for bug fixes / features)

<!--
 Before submitting your PR, please check that your code follows our contribution guidelines:
 https://github.com/Alfresco/alfresco-ng2-components/wiki/Code-contribution-acceptance-criteria
 -->

**What kind of change does this PR introduce?** (check one with "x")

> - [ ] Bugfix
> - [ ] Feature
> - [ ] Code style update (formatting, local variables)
> - [x] Refactoring (no functional changes, no api changes)
> - [ ] Build related changes
> - [ ] Documentation
> - [ ] Other... Please describe:


**What is the current behaviour?** (You can also link to an open issue here)



**What is the new behaviour?**



**Does this PR introduce a breaking change?** (check one with "x")

> - [ ] Yes
> - [x] No


If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

tomgny commented on lib/core/src/lib/app-config/app-config.service.spec.ts:
```diff
@@ -164,24 +164,23 @@ describe('AppConfigService', () => {
-    it('should load external settings', () => {
-        appConfigService.load().then((config) => {
+    it('should load external settings', async () => {
+        await appConfigService.load().then((config) => {
```
do we need to use `.then()` when we are using `await`?

```suggestion
        await appConfigService.load();
        expect(config).toEqual(mockResponse);
```

tomgny commented on lib/core/src/lib/app-config/app-config.service.spec.ts:
```diff
@@ -111,12 +111,12 @@ describe('AppConfigService', () => {
-            expect(property).toBeTruthy();
-        });
-    }));
+        const propertyPromise = appConfigService.select('testProp').toPromise();
```
can we try to use `lastValueFrom() / firstValueFrom()`?
[FE unit test best practices](https://alfresco.atlassian.net/wiki/spaces/TECH/pages/510919005/FE+unit+tests+best+practices#Adapting-a-test-written-with-done()-to-use-promises-instead)

Venrofi replied:
Currently, we're using RxJS v6.6.6 and those functions are supported from v7 onwards 😄

arditdomi reviewed (commented):
Are u sure about this? Keeping fakeAsync with tick I think is fine, but we have to simplify and refactor the tests. For example I see there are nested `fixture.whenStable().then` and we can start by getting rid of those. Do the errors go away if we simply refactor and get rid of the `fixture.whenStable().then`

Venrofi commented:
In most cases, the root cause for the problems with Unhandled Promises was this `fixture.whenStable().then` with fakeAsync. It just didn't work correctly, I guess after the Angular upgrades and all the `expect()` was nested inside of those `.then()`. When I refactored it from `fakeAsync` to normal `async` and simplified the `fixture.whenStable().then` to `await fixture.whenStable()` and got rid of the nesting it worked as intended. Now that the `expect()` is appropriately checked, there are some more failures and plenty of tests with no expectations, but that's a topic for another issue that I've already created a ticket for. I mean you can have a combo of fakeAsync and a normal async inside of it as well because I saw one test that utilises that and it gives you the option to use await but that's complicating rather than simplifying to my mind. 
About the `tick()` - in order for some tests to pass the `tick()` was necessary and when I got rid of `fakeAsync` I tried different methods of substituting the `tick()` and this was the best solution. In some tests the `tick()` was unnecessary but I guess someone did it to be sure they'll pass. I've tested the ones I've touched if this await is really needed and in many cases, I could get away without it 😄

Venrofi commented:
With my changes, there are no more errors regarding Unhandled Promises apart from 3 backend-related tests. I don't know what to do about them really. I think Tom said something about a need to mock a service or something for them.

arditdomi commented:
> In most cases, the root cause for the problems with Unhandled Promises was this `fixture.whenStable().then` with fakeAsync. It just didn't work correctly, I guess after the Angular upgrades and all the `expect()` was nested inside of those `.then()`. When I refactored it from `fakeAsync` to normal `async` and simplified the `fixture.whenStable().then` to `await fixture.whenStable()` and got rid of the nesting it worked as intended. Now that the `expect()` is appropriately checked, there are some more failures and plenty of tests with no expectations, but that's a topic for another issue that I've already created a ticket for. I mean you can have a combo of fakeAsync and a normal async inside of it as well because I saw one test that utilises that and it gives you the option to use await but that's complicating rather than simplifying to my mind. About the `tick()` - in order for some tests to pass the `tick()` was necessary and when I got rid of `fakeAsync` I tried different methods of substituting the `tick()` and this was the best solution. In some tests the `tick()` was unnecessary but I guess someone did it to be sure they'll pass. I've tested the ones I've touched if this await is really needed and in many cases, I could get away without it 😄

https://github.com/Alfresco/alfresco-ng2-components/pull/7832/files#diff-e98fea480e52047f41b7acc01e473d01f6d07657841ecd8cfb29c3ae2580c6c7R417 I don't think this is a good practice. I suggest to keep fakeAsync in combination with tick(); and remove the whenStable() and let's see how many test cases fail without it.

Venrofi commented:
I'll try different scenarios and give you an update 😄

Venrofi commented on lib/core/src/lib/card-view/components/card-view-textitem/card-view-textitem.component.spec.ts:
```diff
@@ -476,123 +474,112 @@ describe('CardViewTextItemComponent', () => {
+        });
 
-        it('should render the error when the editedValue is invalid', fakeAsync((done) => {
+        it('should render the error when the editedValue is invalid', async () => {
```
I've tested a few options on this test and only my solution works correctly. 
![image](https://user-images.githubusercontent.com/64163335/191693779-c9559813-c0ea-4d4d-abc9-8b24daa226e7.png)
![image](https://user-images.githubusercontent.com/64163335/191693841-5b1e95d9-f601-42df-bf4d-8128a5d392f4.png)
![image](https://user-images.githubusercontent.com/64163335/191693873-b1a55fc2-24c1-4663-9e01-747637849361.png)
![image](https://user-images.githubusercontent.com/64163335/191693967-485a38db-0c22-45d6-ab7f-b950a606f1b1.png)
![image](https://user-images.githubusercontent.com/64163335/191694019-e1894bcc-f9ca-405e-8f3b-ed2fb565b7a9.png)

Venrofi commented:
I'm testing different solutions and for some, a remedy is just to delete `.then()`, in some cases using `fakeAsync( async () => ) {} )` gets the job done but I've found one test and I feel there will be more of them where only my solution with `await setTimeoutPromise() ` works. I'll check each test and give you the final update 😄@arditdomi
